addons/survey/controllers/survey_session_manage.py

```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import werkzeug
+from werkzeug.urls import url_join
+
+from odoo import http
+from odoo.http import request
+
+
+class UserInputSession(http.Controller):
+    def _fetch_from_token(self, survey_token):
+        """ Check that given survey_token matches a survey 'access_token'.
+        Unlike the regular survey controller, user trying to access the survey must have full access rights! """
+        return request.env['survey.survey'].search([('access_token', '=', survey_token)])
+
+    @http.route('/survey/session_manage/<string:survey_token>', type='http', auth='user', website=True)
+    def survey_session_manage(self, survey_token, **kwargs):
+        """ Main route used by the host to 'manager' the session.
+        - If the state of the session is 'ready'
+          We render a template allowing the host to showcase the different options of the session
+          and to actually start the session.
+        - If the state of the session is 'in_progress'
+          We render a template allowing the host to show the question results, display the attendee
+          ranking or go to the next question of the session. """
+
+        survey = self._fetch_from_token(survey_token)
+
+        if not survey or survey.session_state == 'not_started':
+            # no open session
+            return werkzeug.utils.redirect('/')
+
+        if survey.session_state == 'ready':
+            base_url = request.env['ir.config_parameter'].sudo().get_param("web.base.url")
+            return request.render('survey.user_input_session_open', {
+                'survey': survey,
+                'survey_url': url_join(base_url, survey.get_start_short_url())
+            })
+        else:
+            template_values = self._prepare_manage_session_values(survey)
+            return request.render('survey.user_input_session_manage', template_values)
+
+    @http.route('/survey/session_next_question/<string:survey_token>', type='json', auth='user', website=True)
+    def survey_session_next_question(self, survey_token, **kwargs):
+        """ This route is called when the host goes to the next question of the session.
+
+        It's not a regular 'request.render' route because we handle the transition between
+        questions using a AJAX call to be able to display a bioutiful fade in/out effect. """
+
+        survey = self._fetch_from_token(survey_token)
+
+        if not survey or survey.session_state == 'not_started':
```

Comment:
The fetch + return "" / False / Forbidden /could be extracted in a _check_and_redirect method (with maybe an optional parameter to differentiate 404 (http) from False (json)). That way we have the logic defined once instead of 4.

Reply:
Each method would still have to return the result of that generic check if it returns something like "error_code_404" or "error_code".
I'm not sure how I could implement that efficiently :(